README.md
Geordi [![Build Status](https://travis-ci.org/makandra/geordi.svg?branch=master)](https://travis-ci.org/makandra/geordi)
======

Geordi is a collection of command line tools we use in our daily work with
Ruby, Rails and Linux at [makandra](http://makandra.com/).

Installation:

    gem install geordi


`geordi`
--------

The `geordi` binary holds most of the utility commands. For the few other
binaries, see the bottom of this file.

You may abbreviate commands by typing only their first letters, e.g. `geordi
con` will boot a development console, `geordi set -t` will setup a project and
run tests afterwards.

For details on commands, e.g. supported options, you may always run
`geordi help <command>`.

### `geordi apache-site VIRTUAL_HOST`

Enable the given virtual host, disabling all others.


### `geordi capistrano COMMAND`

Run a capistrano command on all deploy targets.

Example: `geordi capistrano deploy`


### `geordi chromedriver-update`

Update the chromedriver.

Example: `geordi chromedriver_update`

This command will find and install the matching chromedriver for the currently installed Chrome.


### `geordi clean`

Remove unneeded files from the current directory.


### `geordi commit`

Commit using a story title from Pivotal Tracker.

Example: `geordi commit`

Any extra arguments are forwarded to `git commit -m <message>`.

If there are no staged changes, prints a warning but will continue to create
an empty commit.

On the first execution we ask for your Pivotal Tracker API token. It will be
stored in `~/.config/geordi/global.yml`.


### `geordi console [TARGET]`

Open a Rails console locally or on a Capistrano deploy target.

Open a local Rails console: `geordi console`

Open a Rails console on `staging`: `geordi console staging`


### `geordi cucumber [FILES and OPTIONS]`

Run Cucumber features.

Example: `geordi cucumber features/authentication_feature:3`

Runs Cucumber as you want: with `bundle exec`, using parallel tests, with
a VNC session holding Selenium test browsers, support for using a dedicated
testing firefox and beta support for rerunning failed scenarios.

- *@solo:* Generally, features are run in parallel. However, scenarios tagged
with @solo are excluded and will be run sequentially, _after_ the parallel run.

- *Debugging:* Sometimes, the dot-printing Cucumber formatter does not show
errors. In case a feature fails without a message, try running it with `--debug`
or `-d`.

- *Options:* Any unknown option will be passed through to Cucumber,
e.g. `--format pretty`.

- *VNC sessions:* By default, the test browsers will be run in a VNC session.
This behaviour can be turned off by specifying

    headless_browser: true

in a `.geordi.yml` file in the working directory.


### `geordi delete-dumps [DIRECTORY]`

Delete database dump files (*.dump).

Example: `geordi delete_dumps` or `geordi delete_dumps ~/tmp/dumps`

Recursively search for files ending in `*.dump` and offer to delete those. When
no argument is given, two default directories are searched for dump files: the
current working directory and `~/dumps` (for dumps created with geordi).

Geordi will ask for confirmation before actually deleting files.


### `geordi deploy [STAGE]`

Guided deployment across branches.

Example: `geordi deploy` or `geordi deploy p[roduction]` or `geordi deploy --current-branch`

Merge, push and deploy with a single command! **It always tells what it will do
before it does it.** There are different scenarios where this command is handy:

1) *Production deploy:* From the master branch, run `geordi deploy production`.
   This will merge `master` to `production`, push and deploy to production.

2) *Feature branch deploy:* From a feature branch, run `geordi deploy staging`.
   This will merge the feature branch to `master`, push and deploy to staging.

   To deploy a feature branch directly without merging, run
   `geordi deploy --current-branch`. This feature depends on the environment
   variable `DEPLOY_BRANCH` to be picked up in the respective deploy file.

3) *Simple deploy:* If the source branch matches the target branch, merging will
   be skipped.

Calling the command without arguments will infer the target stage from the
current branch and fall back to master/staging.

Finds available Capistrano stages by their prefix, e.g. `geordi deploy p` will
deploy production, `geordi deploy mak` will deploy a `makandra` stage if there
is a file config/deploy/makandra.rb.

When your project is running Capistrano 3, deployment will use `cap deploy`
instead of `cap deploy:migrations`. You can force using `deploy` by passing the
-M option: `geordi deploy -M staging`.


### `geordi drop-databases`

Interactively delete local databases.

Example: `geordi drop_databases`

Check both MySQL/MariaDB and Postgres on the machine running geordi for databases
and offer to delete them. Excluded are databases that are whitelisted. This comes
in handy when you're keeping your currently active projects in the whitelist files
and perform regular housekeeping with Geordi.

When called with `-P` or `-M` options, only handles Postgres resp. MySQL/MariaDB.

When called with `--postgres <port or local socket>` or `--mysql <port or local socket>`,
will instruct the underlying management commands to use those connection methods
instead of the defaults. This is useful when running multiple installations.

Geordi will ask for confirmation before actually dropping databases and will
offer to edit the whitelist instead.


### `geordi dump [TARGET]`

Handle dumps (see `geordi help dump` for details).

When called without arguments, dumps the development database with `dumple`.

    geordi dump

When called with the `--load` option, sources the specified dump into the
development database.

    geordi dump -l tmp/staging.dump

When called with a capistrano deploy target (e.g. `staging`), remotely dumps
the specified target's database and downloads it to `tmp/`.

    geordi dump staging

When called with a capistrano deploy target and the `--load` option, sources the
dump into the development database after downloading it.

    geordi dump staging -l


### `geordi firefox COMMAND`

Run a command with VNC and test browser set up (alias: chrome).

Example: `geordi firefox b cucumber` or `geordi firefox --setup 24.0`

Useful when you need Firefox for Selenium or the VNC set up, but can't use the
`geordi cucumber` command.

*Install* a special Firefox by calling with `--setup <version>`.

This command is aliased `chrome` for users running Selenium in Chrome.


### `geordi help [COMMAND]`

Describe available commands or one specific command.


### `geordi migrate`

Migrate all databases.

Example: `geordi migrate`

If you are using `parallel_tests`, this runs migrations in your development
environment and `rake parallel:prepare` afterwards. Otherwise, invokes `geordi rake`
with `db:migrate`.


### `geordi png-optimize PATH`

Optimize .png files.

Example: `geordi png-optimize some/directory`

- Removes color profiles: cHRM, sRGB, gAMA, ICC, etc.
- Eliminates unused colors and reduces bit-depth (if possible)
- May reduce PNG file size lossless


### `geordi rake TASK`

Run a rake task in several Rails environments.

Example: `geordi rake db:migrate`

`TASK` is run in the following Rails environments (if present):

- development
- test
- cucumber


### `geordi remove-executable-flags`

Remove executable-flags from files that should not be executable.


### `geordi rspec [FILES]`

Run RSpec.

Example: `geordi rspec spec/models/user_spec.rb:13`

Runs RSpec as you want: with RSpec 1/2 detection, `bundle exec`, rspec_spinner
detection, etc.


### `geordi security-update [STEP]`

Support for performing security updates.

Preparation for security update: `geordi security-update`

Checks out production and pulls.

After performing the update: `geordi security-update finish`

Switches branches, pulls, pushes and deploys as required by our workflow. Tells
what it will do before it does it. In detail:

1. Asks user, if tests are green

2. Pushes production

3. Checks out master and pulls

4. Merges production and pushes in master

5. Deploys staging first, if there is a staging environment

6. Asks user, if deployment log is okay and application is still running on staging

7. Deploys other stages

8. Asks user, if deployment log is okay and application is still running on all other stages

9. Informs user about the next steps


### `geordi server [PORT]`

Start a development server.


### `geordi setup`

Setup a project for the first time.

Example: `geordi setup`

Check out a repository and cd into its directory. Then let `setup` do the tiring
work: run `bundle install`, create `database.yml`, create databases, migrate
(all if applicable).

If a local bin/setup file is found, Geordi skips these steps runs bin/setup
for setup instead.

After setting up, loads a remote database dump into the development db when
called with the `--dump` option:

    geordi setup -d staging

After setting up, runs all tests when called with the `--test` option:

    geordi setup -t


### `geordi shell TARGET`

Open a shell on a Capistrano deploy target.

Example: `geordi shell production`

Lets you select the server to connect to when called with `--select-server`:

    geordi shell production -s


### `geordi tests`

Run all employed tests.


### `geordi unit`

Run Test::Unit.


### `geordi update`

Bring a project up to date.

Example: `geordi update`

Performs: `git pull`, `bundle install` (if necessary) and migrates (if applicable).

After updating, loads a dump into the development db when called with the
`--dump` option:

    geordi update -d staging

After updating, runs all tests when called with the `--test` option:

    geordi update -t

See `geordi help update` for details.


### `geordi version`

Print the current version of geordi.


### `geordi vnc`

Show the hidden VNC window.

Example: `geordi vnc` or `geordi vnc --setup`

Launch a VNC session to the hidden screen where `geordi cucumber` runs Selenium
tests.

When called with `--setup`, will guide through the setup of VNC.


b
---

Runs the given command under `bundle exec` if a `Gemfile` is present in your
working directory. If no `Gemfile` is present just runs the given command:

    b spec spec/models

More information at http://makandracards.com/makandra/684-automatically-run-bundle-exec-if-required


dumple
------

Stores a timestamped database dump for the given Rails environment in `~/dumps`:

    dumple development

More information at http://makandracards.com/makandra/1008-dump-your-database-with-dumple


launchy_browser
---------------

Used by the `geordi cucumber` command. Makes launchy open pages in the user's
browser, as opposed to opening it within the VNC window.


Contributing
============

* Run the tests for the oldest supported ruby version with `bundle exec rake`. Ensure that all other ruby versions in the `travis.yml` pass as well after pushing your feature branch and triggering the travis build.
* Update this `README`. The whole `geordi` section is auto-generated
  by `rake readme`.
* Document your changes in the `CHANGELOG.md` file.

Adding a new command
---------------

Copy `lib/geordi/COMMAND_TEMPLATE` to `lib/geordi/commands/your_command` and
edit it to do what you need it to do. Usually, it is hard to automatedly test
Geordi commands, so make sure you've manually tested it. You can do so by
calling your local geordi like so (adjust paths to your needs):

    # @option -I: add directory to load path
    ruby -I ../geordi/lib ../geordi/bin/geordi <command>
    
    # with debugger
    ruby -r byebug -I ../geordi/lib ../geordi/bin/geordi <command>


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).


## Unreleased

### Compatible changes

### Breaking changes

## 3.2.0 2020-06-24

### Compatible changes

- Add a `.geordi.yml` file to change multiple settings in the project and  `~/.config/geordi/global.yml` for global settings.
- Add #91: Now there is an option to start cucumber without a VNC session. This is configured by the .geordi.yml file.
- Deprecated the `.pt_project_id` file in favor of `.geordi.yml`.
- Deprecated the `~/.gitpt` file in favor of `~/.config/geordi/global.yml`.
- Fix: `|line|.start_with /[A-Z]/` does not work in Ruby < 2.5. This affected `geordi commit`.


## 3.1.0 2020-06-03

### Compatible changes

- Update security-update for improved workflow (#89): Deploy staging first and ask user, if application is still running. Then deploy other stages.


## 3.0.3 2020-05-27

### Compatible changes

- Fix #98: Changing the `config/database.yml` reader from `YAML.load` to `YAML.safe_load` dropped the support for aliases. We now allow aliases and the classes `Time` and `Symbol`. If we encounter further issues with this approach a revert to `YAML.load` would be an option, too.


## 3.0.2 2020-05-18

### Compatible changes

- Fix #95: Method change from `! *.include?` to `*.exclude?` was not valid as we do not have active support in Geordi. Affected commands were `geordi cucumber` and `geordi deploy`.

### Breaking changes


## 3.0.1 2020-05-13

### Compatible changes
- Fix #93: Using `$CHILDSTATUS` instead of `$?` did not work properly. This affected commands like `geordi drop-databases` to fail.
- Fix #92: Geordi fail messages were converted to exceptions by accident. Now they are printed as red error message without the backtrace again.


## 3.0.0 2020-05-04

### Compatible changes

### Breaking changes
- Remove support for Ruby 1.8.7 and Ruby 1.9.3. Bug fixes might still be backported to 2.x, but we will not add any features to 2.x anymore. Please consider to upgrade the Ruby version of your project.


## 2.11.0 2020-05-04

### Compatible changes
- Added the possibility to change the Rails root for the capistrano config via the environment variable `RAILS_ROOT`. This allows you as a gem developer to run a command like `RAILS_ROOT=~/Projects/my-blog geordi console staging` whereas `geordi` uses the capistrano config from `my-blog`. Otherwise you would need to follow the instructions of [this card](https://makandracards.com/makandra/46617-how-to-use-a-local-gem-in-your-gemfile) to test changes in the gem locally.
- Bug fix for "no staged changes" even if there are changes present (#83).
- Fixed deprecation warning for `Thor exit with status 0 on errors` (#84).
- Replaced `Bundler.with_unbundled_env` with `Bundler.with_original_env` (#77). This is a better replacement than 42cd1c4.
- Add deprecation warning `Deprecation warning: Ruby 1.8.7 and 1.9.3 support will be dropped in Geordi 3.x.` to Geordi 2.
- Fix error `thor requires Ruby version >= 2.0.0` for Ruby 1.8.7 and 1.9.3 (https://github.com/makandra/geordi/issues/79#issuecomment-598664191).

### Breaking changes


## 2.10.1 2020-02-17

### Compatible changes
- Add host parameter to mysql dump loader

## 2.10.0 2020-02-11

### Compatible changes
- Fixes [#78](https://github.com/makandra/geordi/pull/78): Add compatibility for more than 9 CPU cores in
  `geordi drop-databases`.


## 2.9.0 2020-01-03

### Compatible changes
- Fixes [#37](https://github.com/makandra/geordi/issues/37): `geordi cucumber` crashes with `--format=pretty`
- Fixes [#27](https://github.com/makandra/geordi/issues/27): Cucumber rerun switch does work when passing files


## 2.8.0 2020-01-02

### Compatible changes
- Fixed [#77](https://github.com/makandra/geordi/issues/77): Remove deprecation warning for "Bundler.clean_system"


## 2.7.0 2019-11-25

### Compatible changes
- Fixed [#68](https://github.com/makandra/geordi/issues/68): The "cucumber" command now fails early when @solo features fail.
- Added: The "setup" command now prints the db adapter when prompting db credentials.
- Fixed [#71](https://github.com/makandra/geordi/issues/71): When used without staged changes, the "commit" command will print a warning and create an empty commit. Any arguments to the command are forwarded to Git.
- Fixed: The "commit" command will not print the extra message any more.
- Added: The "commit" command prints a (progress) loading message. The message is removed once loading is done.


## 2.6.0 2019-11-04

### Compatible changes
- Added [#73](https://github.com/makandra/geordi/issues/73): New command `chromdriver-update` allows you to update your
  chromedriver to the newest matching version.


## 2.5.0 2019-10-25

### Compatible changes
- Added [#69](https://github.com/makandra/geordi/issues/69): If bin/setup exists, prefer it over Geordi's standard setup routine.
- Fixed [#48](https://github.com/makandra/geordi/issues/48): Bundle right before deploying
- Improved [#56](https://github.com/makandra/geordi/issues/56): Print instructions how to set up `DEPLOY_BRANCH` when deploying the current branch.


## 2.4.0 2019-10-21

### Compatible changes
- [#56](https://github.com/makandra/geordi/issues/56): Add current branch option to "deploy" command. This actually sets the `DEPLOY_BRANCH` variable to the current branch during deployment. The deployed application needs to pick up this variable in the respective Capistrano stage file. Example: `set :branch, ENV['DEPLOY_BRANCH'] || 'master'`


## 2.3.0 2019-08-27

### Compatible changes
- Fixes [#70]: Make "console" and "server" commands ready for Rails 6
- Added: CI testing with Travis


## 2.2.0 2019-02-28

### Compatible changes
- Fixes [#67]: Don't run yarn install unless needed


## 2.1.0 2019-02-25

### Compatible changes
- Fixes [#59]: Removedb name prefix when reading whitelist

lib/geordi/gitpt.rb
class Gitpt
  require 'yaml'
  require 'highline'
  require 'tracker_api'

  # This require-style is to prevent Ruby from loading files of a different
  # version of Geordi.
  require File.expand_path('settings', __dir__)

  def initialize
    self.highline = HighLine.new
    self.settings = Geordi::Settings.new
    self.client = build_client
  end

  def run(git_args)
    Geordi::Interaction.warn <<-WARNING unless Geordi::Util.staged_changes?
No staged changes. Will create an empty commit.
    WARNING

    story = choose_story
    if story
      create_commit "[##{story.id}] #{story.name}", *git_args
    end
  end

  private

  attr_accessor :highline, :client, :settings

  def build_client
    TrackerApi::Client.new(token: settings.pivotal_tracker_api_key)
  end

  def load_projects
    project_ids = settings.pivotal_tracker_project_ids
    project_ids.collect { |project_id| client.project(project_id) }
  end

  def applicable_stories
    projects = load_projects
    projects.collect do |project|
      project.stories(filter: 'state:started,finished,rejected')
    end.flatten
  end

  def choose_story
    if Geordi::Util.testing?
      return OpenStruct.new(id: 12, name: 'Test Story')
    end

    loading_message = 'Connecting to Pivotal Tracker ...'
    print(loading_message)
    stories = applicable_stories
    reset_loading_message = "\r#{' ' * (loading_message.length + stories.length)}\r"

    highline.choose do |menu|
      menu.header = 'Choose a story'

      stories.each do |story|
        print '.' # Progress

        state = story.current_state
        owners = story.owners
        owner_is_me = owners.collect(&:id).include?(client.me.id)

        if state == 'started'
          state = HighLine::GREEN + state + HighLine::RESET
        elsif state != 'finished'
          state = HighLine::RED + state + HighLine::RESET
        end

        state += HighLine::BOLD if owner_is_me

        label = "(#{owners.collect(&:name).join(', ')}, #{state}) #{story.name}"
        label = bold(label) if owner_is_me

        menu.choice(label) { return story }
      end

      menu.hidden ''
      print reset_loading_message # Once menu is build
    end

    nil # Return nothing
  end

  def create_commit(message, *git_args)
    extra = highline.ask("\nAdd an optional message").strip
    message << ' - ' << extra if extra != ''

    Geordi::Util.system! 'git', 'commit', '--allow-empty', '-m', message, *git_args
  end

  def bold(string)
    HighLine::BOLD + string + HighLine::RESET
  end

  def highlight(string)
    bold HighLine::BLUE + string
  end

end


lib/geordi/remote.rb
require 'geordi/capistrano_config'
require 'geordi/interaction'
require 'geordi/util'
require 'highline/import'
require 'pathname'

module Geordi
  class Remote

    REMOTE_DUMP_PATH = '~/dumps/dump_for_download.dump'.freeze

    def initialize(stage)
      @stage = stage
      @config = CapistranoConfig.new(stage)
    end

    def select_server
      selected_server = choose do |menu|
        @config.servers.each do |server|
          menu.choice(server) { server }
        end

        # Default to the first listed server (by convention, the first server
        # in the deploy files is the primary one).
        menu.default = '1'
        menu.prompt = 'Connect to? [1] '
      end

      puts
      selected_server
    end

    def dump(options = {})
      # Generate dump on the server
      shell options.merge({
        remote_command: "dumple #{@config.env} --for_download",
        select_server: nil, # Dump must be generated on the primary server
      })

      destination_directory = File.join(@config.root, 'tmp')
      FileUtils.mkdir_p destination_directory
      destination_path = File.join(destination_directory, "#{@stage}.dump")
      relative_destination = Pathname.new(destination_path).relative_path_from Pathname.new(@config.root)

      Interaction.note "Downloading remote dump to #{relative_destination} ..."
      server = @config.primary_server
      Util.system! "scp -C #{@config.user(server)}@#{server}:#{REMOTE_DUMP_PATH} #{destination_path}"

      Interaction.success "Dumped the #{@stage} database to #{relative_destination}."

      destination_path
    end

    def console(options = {})
      shell(options.merge(remote_command: Util.console_command(@config.env)))
    end

    def shell(options = {})
      server_option = options[:select_server]
      server_number = server_option.to_i

      server =  if server_option == 'select_server'
        select_server
      elsif server_number != 0 && server_number <= @config.servers.count
        server_index = server_number - 1
        @config.servers[server_index]
      else
        @config.primary_server
      end

      remote_command = "cd #{@config.remote_root} && #{@config.shell}"
      remote_command << " -c '#{options[:remote_command]}'" if options[:remote_command]

      Interaction.note 'Connecting to ' + server.to_s
      Util.system! 'ssh', "#{@config.user(server)}@#{server}", '-t', remote_command
    end

  end
end


lib/geordi/settings.rb
module Geordi
  class Settings
    require 'yaml'

    require File.expand_path('util', __dir__)
    require File.expand_path('interaction', __dir__)

    GLOBAL_SETTINGS_FILE_NAME = Util.testing? ? './tmp/global_settings.yml'.freeze : File.join(ENV['HOME'], '.config/geordi/global.yml').freeze
    LOCAL_SETTINGS_FILE_NAME = Util.testing? ? './tmp/local_settings.yml'.freeze : './.geordi.yml'.freeze

    ALLOWED_GLOBAL_SETTINGS = %w[ pivotal_tracker_api_key ].freeze
    ALLOWED_LOCAL_SETTINGS = %w[ use_vnc pivotal_tracker_project_ids ].freeze

    def initialize
      read_settings
    end

    # Global settings
    def pivotal_tracker_api_key
      @global_settings['pivotal_tracker_api_key'] || gitpt_api_key_old || inquire_pt_api_key
    end

    def pivotal_tracker_api_key=(value)
      @global_settings['pivotal_tracker_api_key'] = value
      save_global_settings
    end

    # Local settings
    # They should not be changed by geordi to avoid unexpected diffs, therefore
    # there are no setters for these settings
    def use_vnc?
      @local_settings.fetch('use_vnc', true)
    end

    def pivotal_tracker_project_ids
      project_ids = @local_settings['pivotal_tracker_project_ids'] || pt_project_ids_old

      case project_ids
      when Array
        # nothing to do
      when String
        project_ids = project_ids.split(/[\s]+/).map(&:to_i)
      when Integer
        project_ids = [project_ids]
      else
        project_ids = []
      end

      if project_ids.empty?
        puts
        Geordi::Interaction.warn "Sorry, I could not find a project ID in .geordi.yml :("
        puts

        puts "Please put at least one Pivotal Tracker project id into the .geordi.yml file in this directory, e.g."
        puts
        puts "pivotal_tracker_project_ids:"
        puts "- 123456"
        puts
        puts 'You may add multiple IDs.'
        exit 1
      end

      project_ids
    end

    private

    def read_settings
      global_path = GLOBAL_SETTINGS_FILE_NAME
      local_path = LOCAL_SETTINGS_FILE_NAME

      if File.exists?(global_path)
        global_settings = YAML.safe_load(File.read(global_path))
        check_for_invalid_keys(global_settings, ALLOWED_GLOBAL_SETTINGS, global_path)
      end

      if File.exists?(local_path)
        local_settings = YAML.safe_load(File.read(local_path))
        check_for_invalid_keys(local_settings, ALLOWED_LOCAL_SETTINGS, local_path)
      end

      @global_settings = global_settings || {}
      @local_settings = local_settings || {}
    end

    def check_for_invalid_keys(settings, allowed_keys, file)
      invalid_keys = settings.keys - allowed_keys
      unless invalid_keys.empty?
        Geordi::Interaction.warn "Geordi detected unknown keys in #{file}.\n"

        invalid_keys.sort.each do |key|
          puts "* #{key}"
        end

        puts "\nAllowed keys are:"
        allowed_keys.sort.each do |key|
          puts "* #{key}"
        end
        puts

        exit 1
      end
    end

    def save_global_settings
      global_path = GLOBAL_SETTINGS_FILE_NAME
      global_directory = File.dirname(global_path)
      FileUtils.mkdir_p(global_directory) unless File.directory? global_directory
      File.open(global_path, 'w') do |file|
        file.write @global_settings.to_yaml
      end
    end

    # deprecated
    def gitpt_api_key_old
      file_path = File.join(ENV['HOME'], '.gitpt')
      if File.exist?(file_path)
        token = YAML.load_file(file_path).fetch :token
        Geordi::Interaction.warn "The ~/.gitpt file is deprecated."
        Geordi::Interaction.note "The contained setting will be moved to ~/.config/geordi/global.yml."
        Geordi::Interaction.note "If you don't need to work with an older version of geordi you can delete ~/.gitpt now."
        self.pivotal_tracker_api_key = token

        token
      end
    end

    def inquire_pt_api_key
      Geordi::Interaction.warn 'Your settings are missing or invalid.'
      Geordi::Interaction.warn "Please configure your Pivotal Tracker access."
      token = Geordi::Interaction.prompt 'Your API key:'
      self.pivotal_tracker_api_key = token
      puts

      token
    end

    # deprecated
    def pt_project_ids_old
      if File.exist?('.pt_project_id')
        project_ids = File.read('.pt_project_id')
        Geordi::Interaction.warn "The usage of the .pt_project_id file is deprecated."
        Geordi::Interaction.note Util.strip_heredoc(<<-INSTRUCTIONS)
          Please remove this file from your project and add or extend the .geordi.yml file with the following content:
            pivotal_tracker_project_ids: #{project_ids}
        INSTRUCTIONS

        project_ids
      end
    end

  end
end


lib/geordi/cucumber.rb
require 'rubygems'
require 'tempfile'

# This require-style is to prevent Ruby from loading files of a different
# version of Geordi.
require File.expand_path('interaction', __dir__)
require File.expand_path('firefox_for_selenium', __dir__)
require File.expand_path('settings', __dir__)

module Geordi
  class Cucumber

    VNC_DISPLAY = ':17'.freeze
    VNC_SERVER_COMMAND = "vncserver #{VNC_DISPLAY} -localhost -nolisten tcp -SecurityTypes None -geometry 1280x1024".freeze
    VNC_VIEWER_COMMAND = "vncviewer #{VNC_DISPLAY}".freeze
    VNC_ENV_VARIABLES = %w[DISPLAY BROWSER LAUNCHY_BROWSER].freeze

    def run(files, cucumber_options, options = {})
      self.argv = files + cucumber_options.map { |option| option.split('=') }.flatten
      self.settings = GeordiSettings.new

      consolidate_rerun_txt_files
      show_features_to_run
      setup_vnc unless settings.headless_browser

      command = use_parallel_tests?(options) ? parallel_execution_command : serial_execution_command
      Interaction.note_cmd(command) if options[:verbose]

      puts # Make newline
      system command # Util.system! would reset the Firefox PATH
    end

    def launch_vnc_viewer
      fork do
        error = capture_stderr do
          system(VNC_VIEWER_COMMAND)
        end
        unless $?.success?
          if $?.exitstatus == 127
            Interaction.fail 'VNC viewer not found. Install it with `geordi vnc --setup`.'
          else
            Interaction.note 'VNC viewer could not be opened:'
            puts error
            puts
          end
        end
      end
    end

    def restore_env
      VNC_ENV_VARIABLES.each do |variable|
        ENV[variable] = ENV["OUTER_#{variable}"]
      end
    end

    def setup_vnc
      if try_and_start_vnc
        VNC_ENV_VARIABLES.each do |variable|
          ENV["OUTER_#{variable}"] = ENV[variable] if ENV[variable]
        end
        ENV['BROWSER'] = ENV['LAUNCHY_BROWSER'] = File.expand_path('../../bin/launchy_browser', __dir__)
        ENV['DISPLAY'] = VNC_DISPLAY

        Interaction.note 'Run `geordi vnc` to view the Selenium test browsers'
      end
    end

    private

    attr_accessor :argv, :settings

    def serial_execution_command
      format_args = []
      unless argv.include?('--format') || argv.include?('-f')
        format_args = spinner_available? ? ['--format', 'CucumberSpinner::CuriousProgressBarFormatter'] : ['--format', 'progress']
      end
      [use_firefox_for_selenium, 'b', 'cucumber', format_args, escape_shell_args(argv)].flatten.compact.join(' ')
    end

    def parallel_execution_command
      Interaction.note 'Using parallel_tests'
      self.argv = argv - command_line_features

      type_arg = Util.gem_version('parallel_tests') > Gem::Version.new('0.7.0') ? 'cucumber' : 'features'
      features = features_to_run
      features = find_all_features_recursively('features') if features.empty?

      [
        use_firefox_for_selenium,
        'b parallel_test -t ' + type_arg,
        %(-o '#{command_line_options.join(' ')} --tags "#{not_tag('@solo')}"'),
        "-- #{features.join(' ')}",
      ].compact.join(' ')
    end

    def not_tag(name)
      if Util.gem_major_version('cucumber') < 3
        "~#{name}"
      else
        "not #{name}"
      end
    end

    def use_firefox_for_selenium
      path = Geordi::FirefoxForSelenium.path_from_config
      if path
        "PATH=#{path}:$PATH"
      end
    end

    def escape_shell_args(*args)
      args.flatten.collect do |arg|
        arg.gsub(/([\\ "])/) { |_match| "\\#{Regexp.last_match(1)}" }
      end
    end

    def show_features_to_run
      if command_line_options.include? '@solo'
        Interaction.note 'All features tagged with @solo'
      elsif command_line_options.include? 'rerun'
        Interaction.note 'Rerunning failed scenarios'
      elsif features_to_run.empty?
        Interaction.note 'All features in features/'
      else
        notification = 'Only: ' + features_to_run.join(', ')
        notification << ' (from rerun.txt)' if (features_to_run == rerun_txt_features) && (features_to_run != command_line_features)
        Interaction.note notification
      end
    end

    def features_to_run
      @features_to_run ||= begin
        features = find_all_features_recursively(command_line_features)
        features = rerun_txt_features if features.empty?
        features
      end
    end

    def rerun_txt_features
      @rerun_txt_features ||= begin
        if File.exist?('rerun.txt')
          IO.read('rerun.txt').to_s.strip.split(/\s+/)
        else
          []
        end
      end
    end

    def command_line_features
      @command_line_features ||= argv - command_line_options
    end

    def command_line_options
      @command_line_options ||= [].tap do |args|
        # Sorry for this mess. Option parsing doesn't get much prettier.
        argv.each_cons(2) do |a, b|
          break if a == '--' # This is the common no-options-beyond marker

          case a
          when '-f', '--format', '-p', '--profile', '-t', '--tags'
            args << a << b # b is the value for the option
          else
            args << a if a.start_with? '-'
          end
        end

        # Since we're using each_cons(2), the loop above will never process the
        # last arg. Do it manually here.
        last_arg = argv.last
        args << last_arg if (last_arg && last_arg.start_with?('-'))
      end
    end

    def consolidate_rerun_txt_files
      parallel_rerun_files = Dir.glob('parallel_rerun*.txt')
      unless parallel_rerun_files.empty?
        Interaction.note 'Consolidating parallel_rerun.txt files ...'

        rerun_content = []
        parallel_rerun_files.each do |filename|
          rerun_content << File.read(filename).strip
          File.unlink(filename)
        end

        File.open('rerun.txt', 'w') do |f|
          f.puts(rerun_content.join(' '))
        end
      end
    end

    def find_all_features_recursively(files_or_dirs)
      Array(files_or_dirs).map do |file_or_dir|
        if File.directory?(file_or_dir)
          file_or_dir = Dir.glob(File.join(file_or_dir, '**', '*.feature'))
        end
        file_or_dir
      end.flatten.uniq.compact
    end

    def spinner_available?
      @spinner_available ||= File.exist?('Gemfile') && File.open('Gemfile').read.scan(/cucumber_spinner/).any?
    end

    def use_parallel_tests?(options)
      options.fetch(:parallel, true) &&
        features_to_run.size != 1 &&
        Util.gem_available?('parallel_tests') &&
        features_to_run.none? { |f| f.include? ':' }
    end

    def try_and_start_vnc
      # check if vnc is already running
      # return true if vnc_server_running?
      error = capture_stderr do
        system(VNC_SERVER_COMMAND)
      end
      case $?.exitstatus
      when 0,
        98 # was already running after all
        true
      when 127 # not installed
        Interaction.warn 'Could not launch VNC server. Install it with `geordi vnc --setup`.'
        false
      else
        Interaction.warn 'Starting VNC failed:'
        puts error
        puts
        false
      end
    end

    def capture_stderr
      old_stderr = $stderr.dup
      io = Tempfile.new('cuc')
      $stderr.reopen(io)
      yield
      io.rewind
      io.read
    ensure
      io.close
      io.unlink
      $stderr.reopen(old_stderr)
    end

  end
end
